File: Treachery-iOS/Treachery-iOS/Models/Role.swift
//
//  Role.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/12/26.
//

import SwiftUI

enum Role: String, Codable, CaseIterable, Identifiable {
    case leader
    case guardian
    case assassin
    case traitor

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .leader:   return "Leader"
        case .guardian:  return "Guardian"
        case .assassin:  return "Assassin"
        case .traitor:   return "Traitor"
        }
    }

    var color: Color {
        switch self {
        case .leader:   return .mtgLeader
        case .guardian:  return .mtgGuardian
        case .assassin:  return .mtgAssassin
        case .traitor:   return .mtgTraitor
        }
    }

    var winConditionText: String {
        switch self {
        case .leader:
            return "Eliminate all Assassins and Traitors to win."
        case .guardian:
            return "Keep the Leader alive. Eliminate all Assassins and Traitors."
        case .assassin:
            return "Eliminate the Leader while at least one Assassin survives."
        case .traitor:
            return "Be the last player standing."
        }
    }

    /// Role distribution for a given player count.
    /// Returns (leaders, guardians, assassins, traitors).
    static func distribution(forPlayerCount count: Int) -> (leaders: Int, guardians: Int, assassins: Int, traitors: Int) {
        switch count {
        #if DEBUG
        case 1:  return (1, 0, 0, 0)   // Solo: leader only (test card/UI flow)
        case 2:  return (1, 0, 1, 0)   // Leader vs assassin
        case 3:  return (1, 0, 1, 1)   // Leader, assassin, traitor
        #endif
        case 4:  return (1, 0, 2, 1)
        case 5:  return (1, 1, 2, 1)
        case 6:  return (1, 1, 3, 1)
        case 7:  return (1, 2, 3, 1)
        case 8:  return (1, 2, 3, 2)
        default: return (1, 0, 2, 1)
        }
    }

    /// Minimum number of players required to start a game.
    static var minimumPlayerCount: Int {
        #if DEBUG
        return UserDefaults.standard.bool(forKey: "devModeEnabled") ? 1 : 4
        #else
        return 4
        #endif
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/Role.kt
package com.solomon.treachery.model

import androidx.compose.ui.graphics.Color

enum class Role(val value: String) {
    LEADER("leader"),
    GUARDIAN("guardian"),
    ASSASSIN("assassin"),
    TRAITOR("traitor");

    val displayName: String
        get() = when (this) {
            LEADER -> "Leader"
            GUARDIAN -> "Guardian"
            ASSASSIN -> "Assassin"
            TRAITOR -> "Traitor"
        }

    val roleColor: Color
        get() = when (this) {
            LEADER -> Color(0xFFE4C96A)
            GUARDIAN -> Color(0xFF4C8CC9)
            ASSASSIN -> Color(0xFFC94C4C)
            TRAITOR -> Color(0xFF9C4CC9)
        }

    val winConditionText: String
        get() = when (this) {
            LEADER -> "Eliminate all Assassins and Traitors to win."
            GUARDIAN -> "Keep the Leader alive. Eliminate all Assassins and Traitors."
            ASSASSIN -> "Eliminate the Leader while at least one Assassin survives."
            TRAITOR -> "Be the last player standing."
        }

    companion object {
        const val MINIMUM_PLAYER_COUNT = 4

        fun fromValue(value: String): Role? = entries.find { it.value == value }

        fun distribution(playerCount: Int): RoleDistribution = when (playerCount) {
            4 -> RoleDistribution(1, 0, 2, 1)
            5 -> RoleDistribution(1, 1, 2, 1)
            6 -> RoleDistribution(1, 1, 3, 1)
            7 -> RoleDistribution(1, 2, 3, 1)
            8 -> RoleDistribution(1, 2, 3, 2)
            else -> RoleDistribution(1, 0, 2, 1)
        }
    }
}

data class RoleDistribution(
    val leaders: Int,
    val guardians: Int,
    val assassins: Int,
    val traitors: Int
)
